aws/resource_aws_servicecatalog_product.go

```diff
@@ -0,0 +1,293 @@
+	artifactProperties.Name = aws.String(artifactSettings["name"].(string))
+	artifactProperties.Type = aws.String("CLOUD_FORMATION_TEMPLATE")
+
+	url := aws.String(artifactSettings["load_template_from_url"].(string))
+	info := map[string]*string{
+		"LoadTemplateFromURL": url,
+	}
+	artifactProperties.Info = info
+	input.SetProvisioningArtifactParameters(&artifactProperties)
+
+	log.Printf("[DEBUG] Creating Service Catalog Product: %#v %#v", input, artifactProperties)
+	resp, err := conn.CreateProduct(&input)
+	if err != nil {
+		return fmt.Errorf("Creating ServiceCatalog product failed: %s", err.Error())
+	}
+	d.SetId(*resp.ProductViewDetail.ProductViewSummary.ProductId)
+
+	return resourceAwsServiceCatalogProductRead(d, meta)
+}
+
+func resourceAwsServiceCatalogProductRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	input := servicecatalog.DescribeProductAsAdminInput{}
+	input.Id = aws.String(d.Id())
+
+	log.Printf("[DEBUG] Reading Service Catalog Product: %#v", input)
+	resp, err := conn.DescribeProductAsAdmin(&input)
+	if err != nil {
+		if scErr, ok := err.(awserr.Error); ok && scErr.Code() == "ResourceNotFoundException" {
+			log.Printf("[WARN] Service Catalog Product %q not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Reading ServiceCatalog product '%s' failed: %s", *input.Id, err.Error())
+	}
+
+	if err := d.Set("created_time", resp.ProductViewDetail.CreatedTime.Format(time.RFC3339)); err != nil {
+		log.Printf("[DEBUG] Error setting created_time: %s", err)
+	}
+	d.Set("product_arn", resp.ProductViewDetail.ProductARN)
+
+	pvs := resp.ProductViewDetail.ProductViewSummary
+	d.Set("description", pvs.ShortDescription)
+	d.Set("distributor", pvs.Distributor)
+	d.Set("name", pvs.Name)
+	d.Set("owner", pvs.Owner)
+	d.Set("product_type", pvs.Type)
+	d.Set("support_description", pvs.SupportDescription)
+	d.Set("support_email", pvs.SupportEmail)
+	d.Set("support_url", pvs.SupportUrl)
+
+	provisioningArtifactSummary := getProvisioningArtifactSummary(resp)
+	var a []map[string]interface{}
+	artifact := make(map[string]interface{})
+	artifact["description"] = *provisioningArtifactSummary.Description
+	artifact["id"] = *provisioningArtifactSummary.Id
+	artifact["load_template_from_url"] = "only_used_on_initial_create"
+	artifact["name"] = *provisioningArtifactSummary.Name
+	a = append(a, artifact)
+
+	if err := d.Set("artifact", a); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Lookup the first artifact, which was the one created on inital build, by comparing created at time
+func getProvisioningArtifactSummary(resp *servicecatalog.DescribeProductAsAdminOutput) *servicecatalog.ProvisioningArtifactSummary {
+	firstPas := resp.ProvisioningArtifactSummaries[0]
+	for _, pas := range resp.ProvisioningArtifactSummaries {
+		if pas.CreatedTime.UnixNano() < firstPas.CreatedTime.UnixNano() {
+			firstPas = pas
+		}
+	}
+	return firstPas
+}
```

Comment:
@radeksimko 

This is a great question where I would like to get feedback.

When you create a product, it creates an initial product.  This is required.  Subsequent artifacts can be added via create provisioning artifact. 

https://docs.aws.amazon.com/servicecatalog/latest/dg/API_CreateProvisioningArtifact.html

My intention was to create a resource to manage those, and allow the existing resource created by the initial product creation to be managed by the service catalog product.

What are your thoughts on this approach and managing resources that create other resources in situations such as this?

Reply:
So my question/concern is are we _actually_ able to reliably identify the initial artifact? I have seen many AWS APIs in the past which expose an array, but in reality that array is more-or-less randomly sorted on return. This is true for most IAM APIs and others.

What you are suggesting - having a separate `aws_servicecatalog_provisioning_artifact` sounds like a good idea under some circumstances and it would be also helpful because of interpolation - e.g. if you had a list of artifacts in a variable and wanted to iterate over them via `count`.

I was just unable to find an evidence (in docs) that would tell me it's a sorted list and that we can rely on the ordering or that we can otherwise identify the initial artifact.
Also - that's probably more important concern - how is the user going to delete the initial artifact (assuming there is already 1+ other artifacts) in the current setup? It seems the only way would be to destroy & recreate the product which is suboptimal.

We do have an example of resource managing initial "subresource" and delegating more to a stand-alone resource, specifically ALB certs:
https://www.terraform.io/docs/providers/aws/r/lb_listener.html (`certificate_arn`)
https://www.terraform.io/docs/providers/aws/r/lb_listener_certificate.html
However the important fact here is that we _are_ able to tell which certificate is the "initial" - or default one:
https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_Certificate.html (`IsDefault`)

Reply:
@radeksimko Agreed this is not ideal.  Would a better approach be to create a sub resource as part of this resource creation for the initial artifact?  This is a place where I'm not sure on the most idiomatic way to construct the two resources.

Reply:
@radeksimko looking at the API more, it appears that the initial product requires an artifact.  However you will also want to provision more artifacts separately.

My thinking:

* The initial artifact has to be managed with the product.  If you update it, you update it via the product.  It cannot be deleted.
* There is a separate resource for adding new artifacts to an existing product.

The downside is the initial artifact can only be managed as part of the product.

Thoughts?

Reply:
> Would a better approach be to create a sub resource as part of this resource creation for the initial artifact? This is a place where I'm not sure on the most idiomatic way to construct the two resources.

I don't think there's any way to do it at this point. Supporting subresources is something we want to do eventually (it was discussed a number of times within the team), but it's not on our near-term roadmap.

> The initial artifact has to be managed with the product. If you update it, you update it via the product. It cannot be deleted.

The last part would be worrying to me as a user. I think we should allow users to delete that artifact, if they choose to. This leads me to a conclusion that we'll be better off just managing all artifacts inline as I proposed originally.

We might do some refactoring and deprecating when/if Amazon decides to change the API, but at this point it looks like it would be safer to _not_ support artifact as a separate resource (to avoid conflicts between inline artifacts and such resource).

Reply:
yes, that is the idea 👍

Reply:
I mean, structure-wise, naming-wise I think the field should be called `provisioning_artifact` as mentioned elsewhere.

Reply:
@radeksimko I've done a first pass at refactoring as we discussed above.  Can you do a review of the code and let me know if the high level approach matches what we discussed (don't worry about minor code cleanup, I'll do that once we align on the approach).

Another question, the url from where the template was loaded can not be obtained from the API.  This makes it not possible to populate it when importing a product created outside of TF.  What are your thoughts on how to approach importing?  Should it be removed?

Thanks in advance.

Reply:
Actually, hold on that second question, I may have found away.  Let me know your feedback on the overall approach.